annotated.go

```diff
@@ -89,3 +90,268 @@ func (a Annotated) String() string {
 	}
 	return fmt.Sprintf("fx.Annotated{%v}", strings.Join(fields, ", "))
 }
+
+type annotations struct {
+	ParamTags  []string
+	ResultTags []string
+}
+
+// Annotation can be passed to Annotate(f interface{}, anns ...Annotation)
+// for annotating the parameter and result types of a function.
+type Annotation interface {
+	apply(*annotations) error
+	getAnnotatedType(reflect.Type) []reflect.Type
+}
+
+type annotationError struct {
+	target interface{}
+	err    error
+}
+
+func (e *annotationError) Error() string {
+	return e.err.Error()
+}
+
+type paramTags struct {
+	tags []string
+}
+
+func (pt paramTags) apply(ann *annotations) error {
+	if len(ann.ParamTags) > 0 {
+		return fmt.Errorf("cannot apply more than one line of ParamTags")
+	}
+	ann.ParamTags = pt.tags
+	return nil
+}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.In
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (pt paramTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedParams := []reflect.StructField{{
+		Name:      "In",
+		Type:      reflect.TypeOf(In{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumIn(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.In(i),
+		}
+		if i < len(pt.tags) {
+			structField.Tag = reflect.StructTag(pt.tags[i])
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	paramTypes := []reflect.Type{reflect.StructOf(annotatedParams)}
+	return paramTypes
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	p := paramTags{tags}
+	return p
+}
+
+type resultTags struct {
+	tags []string
+}
+
+func (rt resultTags) apply(ann *annotations) error {
+	if len(ann.ResultTags) > 0 {
+		return fmt.Errorf("cannot apply more than one line of ResultTags")
+	}
+	ann.ResultTags = rt.tags
+	return nil
+}
+
+// Given func(T1, T2, T3, ..., TN), this generates a type roughly
+// equivalent to,
+//
+//   struct {
+//     fx.Out
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	errInterface := reflect.TypeOf((*error)(nil)).Elem()
+	numAnnotated := 0
+
+	for i := 0; i < fType.NumOut(); i++ {
+		// guard against error results
+		if fType.Out(i).Implements(errInterface) {
+			continue
+		}
```

Comment:
Instead of Implements, we should maybe look at `==` here. The reason for this is that it is okay for me to do,

```
type MyThing struct{ ... }

func (*MyThing) Error() string

func NewMyThing(...) *MyThing
```

We should ignore it only if it is returned as `error`.

(I suspect we have a test for that somewhere for a different case in dig.)